### ISSUE TITLE
"array size is too large" error based on hardcoded limit -- increase for 64-bit platforms?

### ISSUE THREAD

**Issue by gbuesing** [masa16/narray#46]
If I initialize a very large NArray, I get "ArgumentError: array size is too large", example:

```
> NArray.int(50_000, 50_000)
ArgumentError: array size is too large
```

From following this down into the code in ```narray.c```, I see this is triggered by a check that the total size is > 2147483647 (i.e., 2**31 - 1), which is the maximum positive value for a 32-bit signed binary integer. Ref: http://en.wikipedia.org/wiki/2147483647

On 64-bit platforms, should this size limit be upped to the 64-bit signed integer limit (2**63 -1) ?

**Comment by gbuesing** [comment]
Looked into this a bit more, and it appears that the exception is raised before it hits this check -- 

In na_alloc_struct, the total var is typed to int, which is defined as int32_t. So when total is set to 50_000 * 50_000, it overflows.

 So it looks like the fix would be more complicated than just changing a hardwired number.

**Comment by gbuesing** [comment]
I tried a quick hack -- in na_alloc_struct, I removed the 2147483647 checks, and changed the type of the variables "memsz" and "total" from int to long, and with this version I was able to initialize  NArray.int(50_000, 50_000) without error.

Of course other parts of the code would need to be changed for this narray to actually work correctly. I have no idea if this would be as easy as changing a few more ints to longs, or if it would be much more complicated than that.
